plugins/inputs/ping/ping.go

```diff
@@ -157,204 +141,103 @@ func (p *Ping) Gather(acc telegraf.Accumulator) error {
-				return ""
-			}
-			if len(addrs) > 0 {
-				switch v := addrs[0].(type) {
-				case *net.IPNet:
-					ip = v.IP
-				case *net.IPAddr:
-					ip = v.IP
-				}
-				if len(ip) == 0 {
-					return ""
-				}
-				return ip.String()
-			}
-		}
-	}
-
-	return ""
-}
-
-func hostPinger(binary string, timeout float64, args ...string) (string, error) {
-	bin, err := exec.LookPath(binary)
+func (p *Ping) pingToURLNative(destination string, acc telegraf.Accumulator) {
+	pinger, err := ping.NewPinger(destination)
 	if err != nil {
-		return "", err
-	}
-	c := exec.Command(bin, args...)
-	out, err := internal.CombinedOutputTimeout(c,
-		time.Second*time.Duration(timeout+5))
-	return string(out), err
-}
-
-func filterIPs(addrs []net.IPAddr, filterFunc IsCorrectNetwork) []net.IPAddr {
-	n := 0
-	for _, x := range addrs {
-		if filterFunc(x) {
-			addrs[n] = x
-			n++
-		}
+		p.Log.Errorf("Failed to ping: %v", err)
+		acc.AddError(err)
+		return
 	}
-	return addrs[:n]
-}
-
-func hostResolver(ctx context.Context, ipv6 bool, destination string) (*net.IPAddr, error) {
-	resolver := &net.Resolver{}
-	ips, err := resolver.LookupIPAddr(ctx, destination)
 
-	if err != nil {
-		return nil, err
+	// Required for windows. Despite the method name, this should work without the need to elevate privileges and has been tested on Windows 10
+	if runtime.GOOS == "windows" {
+		pinger.SetPrivileged(true)
 	}
 
-	if ipv6 {
-		ips = filterIPs(ips, isV6)
+	// The interval cannot be below 0.2 seconds, matching ping implementation: https://linux.die.net/man/8/ping
+	if p.PingInterval < 0.2 {
+		pinger.Interval = time.Duration(.2 * float64(time.Second))
 	} else {
-		ips = filterIPs(ips, isV4)
-	}
-
-	if len(ips) == 0 {
-		return nil, errors.New("Cannot resolve ip address")
-	}
-	return &ips[0], err
-}
-
-func isV4(ip net.IPAddr) bool {
-	return ip.IP.To4() != nil
-}
-
-func isV6(ip net.IPAddr) bool {
-	return !isV4(ip)
-}
-
-func (p *Ping) pingToURLNative(destination string, acc telegraf.Accumulator) {
-	ctx := context.Background()
-	interval := p.PingInterval
-	if interval < 0.2 {
-		interval = 0.2
+		pinger.Interval = time.Duration(p.PingInterval * float64(time.Second))
 	}
 
-	timeout := p.Timeout
-	if timeout == 0 {
-		timeout = 5
+	// If no timeout is given default to 5 seconds, matching original implementation
+	if p.Timeout == 0 {
+		pinger.Timeout = time.Duration(5) * time.Second
+	} else {
+		pinger.Timeout = time.Duration(p.Timeout) * time.Second
 	}
 
-	tick := time.NewTicker(time.Duration(interval * float64(time.Second)))
-	defer tick.Stop()
-
-	if p.Deadline > 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, time.Duration(p.Deadline)*time.Second)
-		defer cancel()
-	}
+	go func() {
+		if p.Deadline <= 0 {
+			return
+		}
+		// If deadline is set ping exits regardless of how many packets have been sent or received
+		timer := time.AfterFunc(time.Duration(p.Deadline)*time.Second, func() {
+			pinger.Stop()
+		})
+		defer timer.Stop()
+	}()
```

Comment:
AfterFunc returns immediately, this is going to stop the timer immediately after it starts.
Looks like you don't need a goroutine for this.
```suggestion
	if p.Deadline > 0 {
		// If deadline is set ping exits regardless of how many packets have been sent or received
		timer := time.AfterFunc(time.Duration(p.Deadline)*time.Second, func() {
			pinger.Stop()
		})
	}
```

Reply:
That looks a lot better, thanks!